src/amo/withExperiment.js

```diff
@@ -175,15 +181,79 @@ export const withExperiment =
         WrappedComponent,
       )})`;
 
-      componentDidMount() {
-        const { _getVariant, _isExperimentEnabled, cookies, id, variants } =
-          this.props;
-
+      initialExperimentSetup(props) {
+        const {
+          _getVariant,
+          _isExperimentEnabled,
+          dispatch,
+          id,
+          storedVariant,
+          variants,
+        } = props;
+
+        let { variant } = this;
         const isEnabled = _isExperimentEnabled({ _config, id });
         const registeredExperiments = this.getExperiments();
         const experimentInCookie = this.cookieIncludesExperiment(
           registeredExperiments,
         );
+        const addExperimentToCookie = !experimentInCookie;
+        const variantFromStore = storedVariant;
+
+        if (isEnabled && !variant) {
+          // Use the variant in the cookie if one exists, otherwise use the
+          // variant from the Redux store.
+          if (experimentInCookie) {
+            variant = registeredExperiments[id];
+          } else if (variantFromStore) {
+            variant = variantFromStore.variant;
+          }
+
+          // Do we need to store the variant in the cookie?
+          if (addExperimentToCookie) {
+            // Determine the variant if we don't already have one.
+            variant = variant || _getVariant({ variants }).id;
+
+            // If we are on the server, we need to store the variant in the Redux
+            // store.
+            if (_config.get('server')) {
+              const variantToStore: StoredVariant = {
+                id,
+                variant,
+              };
+              dispatch(
+                storeExperimentVariant({ storedVariant: variantToStore }),
+              );
+            }
+          }
+        }
+
+        return {
+          // We only need to add the experiment to the cookie if we have a
+          // variant.
+          addExperimentToCookie: addExperimentToCookie && variant,
+          registeredExperiments,
+          variant,
+          variantFromStore,
+        };
+      }
+
+      constructor(props: withExperimentInternalProps) {
```

Comment:
I suggest moving the constructor up because I was confused when I read `let { variant } = this;` above.